keep-libruntime/src/workload.rs

```diff
@@ -0,0 +1,257 @@
+        modules: HashMap<String, Box<[u8]>>,
+        function_module: String,
+        function_name: String,
+        function_args: Box<[String]>,
+    ) -> Self {
+        Self {
+            modules,
+            function_module,
+            function_name,
+            function_args,
+        }
+    }
+
+    /// Retrieves a serliazed WASM workload from a Reader.
+    pub fn deserialize_from_reader<R: std::io::Read>(r: R) -> Result<Self> {
+        serde_json::from_reader(r).map_err(Into::into)
+    }
+
+    /// Runs an initialized workload, allowing to override the parameter with the given options
+    pub fn run(
+        &self,
+        function_module: Option<&str>,
+        function_name: Option<&str>,
+        function_args: Option<&[Val]>,
+    ) -> Result<Box<[Val]>> {
+        let function_name = function_name.unwrap_or(&self.function_name);
+
+        // Wasmtime API and helper variables
+        let engine = wasmtime::Engine::default();
+        let store = wasmtime::Store::new(&engine);
+        let mut module_registry: HashMap<String, wasmtime::Instance> = HashMap::new();
+
+        // Instantiate WASI
+        let wasi = wasmtime_wasi::instantiate_wasi(
+            // preopened_dirs: &[(String, File)],
+            Default::default(),
+            // argv: &[String],
+            Default::default(),
+            // environ: &[(String, String)],
+            Default::default(),
+        )
+        .expect("couldn't instantiate WASI");
+        module_registry.insert(
+            "wasi_unstable".to_string(),
+            wasmtime::Instance::from_handle(&store, wasi),
+        );
+
+        let instance = {
+            // Resolve the app_module imports using the module_registry.
+            // TODO: process all workload data
+            let module_name = function_module.unwrap_or(&self.function_module);
+            let app_module = wasmtime::Module::new_with_name(
+                &store,
+                &self
+                    .modules
+                    .get(module_name)
+                    .ok_or_else(|| "workload data is empty".to_string())?,
+                module_name,
+            )?;
+            let imports = app_module
+                .imports()
+                .iter()
+                .map(|import| {
+                    let module_name = import.module();
+                    if let Some(instance) = module_registry.get(module_name) {
+                        let field_name = import.name();
+                        if let Some(export) = instance.get_export(field_name) {
+                            Ok(export.clone())
+                        } else {
+                            Err(format!(
+                                "Import {} was not found in module {}",
+                                field_name, module_name
+                            ))
+                        }
+                    } else {
+                        Err(format!("Import module {} was not found", module_name))
```

Comment:
String errors are not very good for libraries. Ok, there is not much action to do in the error case.. Just a thought

Reply:
I wonder where this error will be propagated to. If a user is supplying the workload synchronously it's easy to display it, but if there's an orchestrator involved it's not that obvious.